### DISCUSSION TITLE
Parse mark improvements

### DISCUSSION THREAD

**Discussion by TPGamesNL** [SkriptLang/Skript#4177] [Suggestions] [answered]
Currently, the parse marks require the `¦` character, but this isn't great because it's not in ASCII. Over at skript-parser, they chose the character `:` to be used for parse marks. This could solve some encoding issues for addons, as these characters are often the cause of this. https://github.com/SkriptLang/Skript/pull/4176#discussion_r667341082

One other issue with parse marks, is that they can be difficult to use. This could be solved by changing parse marks into parse tags: strings, not integers. 
For example (these examples use `¦` instead of `:`): 
- `(boots¦boots|leggings¦leggings|chestplate¦chestplate|helmet¦helmet) (item|slot¦slot)` (with code `parseResult.tags.contains("slot")`)

- instead of `(0¦boots|1¦leggings|2¦chestplate|3¦helmet) (item|4¦slot)` (with code `parseResult.mark & 4 != 0`)

- additionally, we could add the special tag name `*`, which would be equal to the content of that option, simplifying the pattern to `(*¦boots|*¦leggings|*¦chestplate|*¦helmet) (item|*¦slot)` (same code: `parseResult.tags.contains("slot")`)

This system could easily be made backwards compatible for parse marks.

**Comment by APickledWalrus** [comment]
I think this idea would be a great improvement to the current system. What do you think about making this only available for colons? If we were to switch over to then completely, it would help with that (in getting devs to switch over). It would also make it simple to handle backwards compatibility, as you wouldn't have to worry about potentially parsing differently. 

ParseResult#mark could maybe be deprecated too and have a comment detailing/linking to the new system.

**Reply by TPGamesNL** [reply]
I think that's a great idea :D

**Comment by Mwexim** [comment]
I concur, and I especially like the usage of strings instead of integers as parse marks. Perhaps you can even do something like:
- `*:(boots|leggings|chestplate|helmet)` to automatically give each part of the group its own special tag name
- `:(boots|leggings|chestplate|helmet)` to automatically give each part of the group its own numerical value starting from 0
These patterns look ugly, of course, but they would be quite useful. Make sure to implement the '*' symbol in such a way that it doesn't clash with syntaxes like the arithmetic operators.

**Reply by abandonedaccount6235** [reply]
another idea: `(:boots|:leggings|:chestplate|:helmet) (item|:slot)` instead of `(*:boots|*:leggings|*:chestplate|*:helmet) (item|*:slot)`. In this syntax, the absence of a tag name implies that the content should be used as the name. Named regex groups from Python might be a good source of inspiration https://docs.python.org/3/howto/regex.html#non-capturing-and-named-groups

**Comment by Mwexim** [comment]
For the record, I have made a [pull request](https://github.com/SkriptLang/skript-parser/pull/116) over at skript-parser which adds this functionality, but preserves the integer parse marks by having a utility method that parses all numerical values.

This is probably a way that still comforts Skript's need to support addons.

**Comment by TPGamesNL** [comment, marked as answer]
I've implemented this in #4176, and explained some stuff here: https://github.com/SkriptLang/Skript/pull/4176#issuecomment-882056536
If anything else should be added to this, lmk
